Cherry-picking virtual dependencies

alalazo opened the pull request:
depends on #39508 
fixes #15443

This PR makes it possible to select only a subset of virtual dependencies from a spec that _may_ provide more. To select providers, a syntax to specify edge attributes is introduced:
```
hdf5 ^[virtuals=mpi] mpich
```
With that syntax we can concretize specs like:
```console
$ spack spec strumpack ^[virtuals=mpi] intel-parallel-studio+mkl ^[virtuals=lapack] openblas
```
and get this concrete DAG:

![strumpack_virtuals](https://user-images.githubusercontent.com/4199709/216620232-749c02b5-c81a-4de2-8949-21d00bc8c49f.svg)


On `develop` this would currently fail with:
```console
$ spack spec strumpack ^intel-parallel-studio+mkl ^openblas
==> Error: Spec cannot include multiple providers for virtual 'blas'
    Requested 'intel-parallel-studio' and 'openblas'
```

In package recipes, virtual specs that are declared in the same `provides` directive need to be provided _together_. This means that e.g. `openblas`, which has:
```python
provides("blas", "lapack")
```
needs to provide both `lapack` and `blas` when requested to provide at least one of them.

## Additional notes

This capability is needed to model compilers. Assuming that languages are treated like virtual dependencies, we might want e.g. to use LLVM to compile C/C++ and Gnu GCC to compile Fortran. This can be accomplished by the following[^1]:
```
hdf5 ^[virtuals=c,cxx] llvm ^[virtuals=fortran] gcc
```

[^1]: We plan to add some syntactic sugar around this syntax, and reuse the `%` sigil to avoid having a lot of boilerplate around compilers.

Modifications:
- [x] Add syntax to interact with edge attributes from spec literals
- [x] Add concretization logic to be able to cherry-pick virtual dependencies
- [x] Extend semantic of the `provides` directive to express when virtuals need to be provided together
- [x] Add unit-tests and documentation

alalazo committed:
Allow Spack to parse edge attributes from cli

alalazo committed:
Allow selecting a provider for a virtual from cli

alalazo committed:
Add semantic for providers that need to go together

alalazo committed:
Add documentation

alalazo commented:
@spackbot run pipeline

RMeli commented:
Hello. I tried this PR with the following spec:

```bash
cp2k@master +dlaf +libxc +libint +spglib build_system=cmake ^[virtuals=blas,lapack,scalapack,mkl]
    intel-mkl@2020.4.304 threads=openmp ^fftw ^cray-mpich
```

It returned this error, which is very cryptic:
```text
==> Error: 
        cp2k@master+dlaf+libint+libxc+spglib build_system=cmake ^cray-mpich ^fftw ^[virtuals=blas,lapack,mkl,scalapack] intel-mkl@2020.4.304 threads=openmp is unsatisfiable, errors are:

```

That is, no error messages are actually printed for `errors are:`.

However, by removing `mkl` I was able to concretise the environment with both `intel-mkl` and `fftw` and everything seems to works as expected:
```bash
cp2k@master +dlaf +libxc +libint +spglib build_system=cmake ^[virtuals=blas,lapack,scalapack]
    intel-mkl@2020.4.304 threads=openmp ^fftw ^cray-mpich
```

This is a really neat and useful feature.

becker33 commented on lib/spack/spack/parser.py:
```diff
@@ -298,6 +307,18 @@ def next_spec(
+                "or a node attribute (version, variant, etc.)"
+            )
+            raise SpecParsingError(msg, self.ctx.current_token, self.literal_str)
+        if root_spec.concrete:
+            raise spack.spec.RedundantSpecError(root_spec, "^" + str(dependency))
```
Isn't this condition impossible now that we defer evaluating parsed hashes until concretization time?

alalazo replied:
I think so, but note that this was copied verbatim from our current develop. I would rather solve this in another PR, where we can discuss if we want `/abc ^zlib` to be valid or not.

alalazo replied:
@becker33 See #39574

becker33 commented on lib/spack/spack/parser.py:
```diff
@@ -278,7 +281,13 @@ def next_spec(
+                edge_properties.setdefault("deptypes", ())
+                edge_properties.setdefault("virtuals", ())
+                root_spec._add_dependency(dependency, **edge_properties)
+            elif self.ctx.accept(TokenType.DEPENDENCY):
```
I think this could be rewritten for more code reuse if you first check for either `DEPENDENCY` or `START_EDGE_PROPERTIES`.

```suggestion
            if self.ctx.accept(TokenType.Dependency) or self.ctx.accept(TokenType.START_EDGE_PROPERTIES):
                edge_properties = {}
                if self.ctx.token.kind == TokenType.START_EDGE_PROPERTIES:
                    edge_properties = EdgeAttributeParser(self.ctx, self.literal_str).parse()
                edge_properties.setdefault("deptypes", ())
                edge_properties.setdefault("virtuals", ())
```
And then the line 294/303 call to `_add_dependency` would change to:
```
root_spec._add_dependency(dependency, **edge_properties)
```

And then the entire `_parse_node` method that you had to add becomes redundant.

alalazo replied:
I just reused `_parse_node` here (forgot to do it before)

becker33 commented on lib/spack/spack/solver/concretize.lp:
```diff
@@ -293,18 +293,42 @@ error(1, Msg) :- attr("node", Package),
+     condition_holds(ID),
+     provided_together(ID, SetID, Package, Virtual1),
+     provided_together(ID, SetID, Package, Virtual2),
+     Virtual1 != Virtual2.
```
Is this important for correctness or is it just breaking symmetry for performance?

alalazo replied:
It's important for correctness. If 2 or more variants are provided together, and one of them isn't "provided" it's an error.

alalazo replied:
We can probably find smarter ways to write this rule, if this starts to be an issue for performance

becker33 commented on lib/spack/spack/solver/concretize.lp:
```diff
@@ -323,32 +355,37 @@ error(100, "Cannot select a single provider for virtual '{0}'", Virtual)
+  :- attr("virtual_node", Virtual),
+     provider(P1, Virtual),
+     provider(P2, Virtual),
+     P1 < P2.
```
Didn't we decide that cardinality rules like below are more performant?

```suggestion
     2 { provider(Provider, Virtual) }.
```

alalazo replied:
Outdated by rebase

becker33 commented on lib/spack/spack/solver/concretize.lp:
```diff
@@ -323,32 +355,37 @@ error(100, "Cannot select a single provider for virtual '{0}'", Virtual)
 % asks to install "mpich" and "hdf5+mpi" and we want "mpich" to
 % be the mpi provider)
-provider(Package, Virtual) :- attr("node", Package), virtual_condition_holds(Package, Virtual).
+provider(Package, Virtual) :- attr("root", Package), virtual_condition_holds(Package, Virtual).
```
This seems to imply that we may need a virtual-selection mechanism in unified environments someday. But that's hopefully out of scope for this PR.

alalazo replied:
Outdated by rebase

becker33 commented on lib/spack/spack/solver/concretize.lp:
```diff
@@ -323,32 +355,37 @@ error(100, "Cannot select a single provider for virtual '{0}'", Virtual)
-% A package cannot be the actual provider for a virtual if it does not
-% fulfill the conditions to provide that virtual
-:- provider(Package, Virtual), not virtual_condition_holds(Package, Virtual),
-   internal_error("Virtual when provides not respected").
```
This is still true, right? Just not necessary for correctness?

If so, I think it's probably better to leave it in -- IIUC integrity constraints like this will improve performance, because it can short-circuit out of directions that will never work.

alalazo replied:
Oudated by rebase

becker33 commented on lib/spack/spack/spec.py:
```diff
@@ -4135,11 +4151,17 @@ def __getitem__(self, name):
+                        for x in order()
+                        if (not x.virtual)
+                        and any(name in edge.virtuals for edge in x.edges_from_dependents())
+                    ),
```
Shouldn't the line after this be unreachable, assuming we've properly constructed virtuals on edges for old specs that we read in?

becker33 commented on lib/spack/spack/spec.py:
```diff
@@ -147,14 +147,14 @@
 #: Default format for Spec.format(). This format can be round-tripped, so that:
 #:     Spec(Spec("string").format()) == Spec("string)"
-default_format = (
+DEFAULT_FORMAT = (
```
Is there a reason for this change? It introduces a bunch of orthogonal changes into the diff for this PR

alalazo replied:
I'll extract that. The reason is to make apparent where we are using global constants in the code.

alalazo replied:
See #39573

becker33 commented on lib/spack/spack/test/build_environment.py:
```diff
@@ -541,3 +541,13 @@ def test_setting_attributes(self, default_mock_concretization):
             assert current_module.SOME_ATTRIBUTE == 1
+
+
+def test_monkey_patching_works_across_virtual(default_mock_concretization):
```
Why not use monkeypatch if the purpose is to test monkeypatch?

alalazo replied:
Because we want to test monkey-patching a package like a user would do, not with an internal mechanism used by pytest.

becker33 commented on var/spack/repos/builtin/packages/strumpack/package.py:
```diff
@@ -67,7 +67,7 @@ class Strumpack(CMakePackage, CudaPackage, ROCmPackage):
     depends_on("blas")
     depends_on("lapack")
-    depends_on("openblas threads=openmp", when="^openblas")
+    depends_on("openblas threads=openmp", when="^[virtuals=lapack,blas] openblas")
```
I don't follow from the concretizer changes why this is necessary, and it concerns me a bit. Does it matter which virtuals openblas is providing whether this requirement holds?

This brings to mind a related worry -- if `foo` depends unconditionally on `openblas` (not through the virtual interface, and `bar` depends on `foo` and `blas`, does this implementation prevent `bar ^[virtuals=blas] intel-parallel-studio`? I would think it should (since `foo` is linking `libopenblas.so`).

alalazo replied:
> I don't follow from the concretizer changes why this is necessary, and it concerns me a bit. Does it matter which virtuals openblas is providing whether this requirement holds?

It is not necessary, but having or not the virtuals has an effect. I added a proper test in https://github.com/spack/spack/pull/35322/commits/bfce9f8add86c667f34ef2533a12aa77e775626a to check that virtuals on edges are accounted for in conditions.

In any case, I can revert this specific change (probably this went in accidentally while I was developing).

tgamblin replied:
What's the effect here?  Is this affecting default concretizations?

alalazo replied:
Reverted in https://github.com/spack/spack/pull/35322/commits/56c362f65b397b64c3b98b50548806a963a8d622

This was to test that an edge attribute matters if part of a condition, but there is also a unit test to ensure this semantic works as intended.

alalazo replied:
To be clear, this would matter in cases where we need to specialize on a dependency that could provide more than one virtual. For instance:
```python
depends_on("foo+intel", when="^intel-parallel-studio")
```
would trigger the dependency on `foo+bar` if `intel-parallel-studio` is a dependency too. While:
```python
depends_on("foo+intel", when="^[virtuals=mpi] intel-parallel-studio")
```
is more specific and would trigger the dependency only if `intel-parallel-studio` is the `mpi` provider.

becker33 commented on lib/spack/spack/test/spec_semantics.py:
```diff
@@ -1003,6 +1003,103 @@ def test_spec_override(self):
+        [
+            # openblas-with-lapack needs to provide blas and lapack together
+            "netlib-scalapack ^[virtuals=blas] intel-parallel-studio ^openblas-with-lapack",
+            # intel-* provides blas and lapack together, openblas can provide blas only
```
Isn't this description reversed?

alalazo replied:
No, for some historical reason I don't have an explanation for, in our mock repository `openblas` provides blas only and we have `openblas-with-lapack` that provides both together.

becker33 commented on lib/spack/spack/test/spec_semantics.py:
```diff
@@ -1003,6 +1003,103 @@ def test_spec_override(self):
+            # Test that we can mix dependencies that provide an overlapping
+            # sets of virtual dependencies
+            (
+                "netlib-scalapack ^[virtuals=mpi] intel-parallel-studio "
+                "^[virtuals=lapack] openblas-with-lapack",
```
Assuming `netlib-scalapack ^intel-parallel-studio ^openblas-with-lapack` would generate an error, can we add a test confirming that?

alalazo replied:
It doesn't generate an error. Similarly to other spec properties, clingo will try to respect the request, and will place virtuals on edges accordingly. This is a real example with:
```console
$ spack graph -d --color strumpack ^openblas ^intel-parallel-studio > strumpack.dot
```
which gives back:
![strumpack](https://github.com/spack/spack/assets/4199709/e5a45840-ec0a-4c07-81c6-ded9e7cb5ae5)

alalazo committed:
Only print errors if there are errors to print

haampie reviewed (dismissed):
Mostly a bike shedding comment:

I would reserve `[` and `]` for other features. Cause (a) it could be used to express constraints that aren't flat `a ^[b ^[c ^d]]`, (b) other obvious brackets need escaping (c) you can do the same with just one character instead of two

So I'd suggest ```pkg_a ^ :type=link :virtual=mpi pkg_b```. (Or maybe even `pkg_a ^pkg_b :virtuals=mpi ordinary_variant=x`)

alalazo commented:
> I would reserve [ and ] for other features.

Whatever we choose, either `[` and `]` or `:`, I would use that as a delimiter and group edge attributes immediately after `^`. This would slightly simplify the parser but more importantly it wouldn't allow writing things like:
```
pkg_a ^openblas +foo :+bar fee=baz :baz=fu 
```
where a spec continuously interchange between edge and node variant. Also, we don't need to give meaning to expressions like:
```
pkg_a :virtual=mpi
```
which might be allowed with a single character.

alalazo committed:
Spec.satisfies: check edge properties

alalazo committed:
Spec.intersects: relax semantic on providers

alalazo committed:
Remove dead code (functions are not used anymore)

alalazo commented:
@spackbot run pipeline

alalazo committed:
Fixup after rebase

tgamblin commented on lib/spack/docs/basic_usage.rst:
```diff
@@ -1525,6 +1525,29 @@ any MPI implementation will do.  If another package depends on
+might want to utilize just a subset of what they could provide, and use other providers for virtuals they
+need.
+
+With Spack it is possible to bind virtual dependencies explicitly using the ``virtuals`` edge attribute:
```
I don't think we should mention that this is about edge attributes here, unless we're going to define edge attribute and link to it somewhere:

```
It is possible to be more explicit and tell Spack which dependency should provide which virtual, using a special syntax:
```

alalazo replied:
Done in https://github.com/spack/spack/pull/35322/commits/f2dd4b40906b9625abd0de0dfbc8cf37f4217196

tgamblin commented on lib/spack/docs/packaging_guide.rst:
```diff
@@ -2937,6 +2937,34 @@ The ``provides("mpi")`` call tells Spack that the ``mpich`` package
+called ``libopenblas``. A package needing ``lapack`` and ``blas`` can either use ``openblas`` as a provider of both, or
+not use ``openblas`` at all.
+
+To express this constraint in a package, the two virtual dependencies must be listed in the same ``provides`` directive:
```
```
well-known example is ``openblas``, which provides both the ``lapack`` and ``blas`` API in a single ``libopenblas`` library. A package that needs ``lapack`` and ``blas`` must either use ``openblas``
to provide both, or not use ``openblas`` at all. It cannot pick one or the other.

To express this constraint in a package, the two virtual dependencies must be listed in the same ``provides`` directive:
```

alalazo replied:
Done in https://github.com/spack/spack/pull/35322/commits/f2dd4b40906b9625abd0de0dfbc8cf37f4217196

tgamblin commented on lib/spack/docs/packaging_guide.rst:
```diff
@@ -2937,6 +2937,34 @@ The ``provides("mpi")`` call tells Spack that the ``mpich`` package
+
+This makes it impossible to select ``openblas`` as a provider for one of the two
+virtual dependencies and not for the other. Any request to do so results in an
+error message:
```
If you try to, Spack will report an error:

alalazo replied:
Done in https://github.com/spack/spack/pull/35322/commits/f2dd4b40906b9625abd0de0dfbc8cf37f4217196

alalazo commented:
Hmm, I was able to concretize the environment on Lassen

alalazo committed:
Fix luajit providers

With the new semantic, lujit and lualang must be
provided together by these specs.

tgamblin commented on lib/spack/spack/parser.py:
```diff
@@ -6,15 +6,16 @@
 
-    node_options  = [@(version_list|version_pair)] [%compiler] { variant }
+    node_options    = [@(version_list|version_pair)] [%compiler] { variant }
+    edge_properties = [[ { bool_variant | key_value } ]]
```
you've got the `[]` specified twice -- once above under `spec` and once here.  I'd just leave them under `spec` and do single `[]` here.

alalazo replied:
Done in https://github.com/spack/spack/pull/35322/commits/ffd8f7d60f0d6d7e129908a979c650969f265e8b

tgamblin commented on lib/spack/spack/parser.py:
```diff
@@ -278,23 +282,34 @@ def next_spec(
+            msg = (
+                "this dependency sigil needs to be followed by a package name "
+                "or a node attribute (version, variant, etc.)"
+            )
```
The error message points to the wrong character when triggered this way:

```
> spack spec hdf5 ^[virtuals=mpi]
==> Error: this dependency sigil needs to be followed by a package name or a node attribute (version, variant, etc.)
hdf5 ^[virtuals=mpi]
                   ^
```

I think the answer is to reformulate the message to be more clear, e.g. "the dependency sigil and any optional edge attributes must be followed by a package name"

alalazo replied:
Done in https://github.com/spack/spack/pull/35322/commits/84a310178460f45bf70134906f67e7ee1fda6f90

tgamblin commented on lib/spack/spack/parser.py:
```diff
@@ -432,6 +447,41 @@ def parse(self, initial_spec: spack.spec.Spec) -> spack.spec.Spec:
+                name = name.strip("'\" ")
+                value = value.strip("'\" ").split(",")
+                attributes[name] = value
+                if name not in ("deptypes", "virtuals"):
```
I am not sure these are the best names -- renaming them will change the spec format, though, right?  I think it would be better for the user to specify `type=...` and `virtual=...`... Is it worth implementing that at this point before this becomes cemented?

With use variants, we'll have more usable names for edge attributes... but these are always going to be reserved as internal ones.  We *could* implement `type` and `virtual` as aliases and reserve `type`, `deptypes`, `virtual`, and `virtuals`, w/o changing the schema.  These are unlikely to be used for anything else.

alalazo replied:
> renaming them will change the spec format, though, right?

Yes, since in the JSON representation these are basically reserved edge variant names.

> Is it worth implementing that at this point before this becomes cemented?

I would say no, but in any case I would do it in a separate PR (to come shortly after) so we can keep the two modifications (feature and renaming) separated.

> We could implement type and virtual as aliases and reserve type, deptypes, virtual, and virtuals, w/o changing the schema. These are unlikely to be used for anything else.

Yes, this could be done.

tgamblin commented on lib/spack/spack/spec.py:
```diff
@@ -1463,6 +1463,26 @@ def edges_to_dependencies(self, name=None, depflag: dt.DepFlag = dt.ALL):
+        for edge in edges:
+            union.update_deptypes(edge.depflag)
+            union.update_virtuals(edge.virtuals)
+        deptypes_str = (
+            f"deptypes={','.join(dt.flag_to_tuple(union.depflag))}" if union.depflag else ""
+        )
+        virtuals_str = f"virtuals={','.join(union.virtuals)}" if union.virtuals else ""
+        if not deptypes_str and not virtuals_str:
+            return ""
+        result = f"{deptypes_str} {virtuals_str}".strip()
+        return f"[{result}]"
```
```suggestion
        attrs = " ".join(
            f"deptypes={','.join(dt.flag_to_tuple(union.depflag))}" if union.depflag else "",
            f"virtuals={','.join(union.virtuals)}" if union.virtuals else "",
        )
        return f"[{attrs}]" if attrs else ""
```

becker33 replied:
I don't think this suggestion is quite correct. It looks to me like it would print `[" "]` if nothing needs to be printed, since using `" ".join` on two empty strings will create the non-empty string with a single space.

tgamblin replied:
Add `.strip()` then?  I think it's still clearer.

tgamblin reviewed (changes requested):
Not completely done, but here is a partial review.

alalazo committed:
Improved documentation according to review

alalazo committed:
Fix grammar

alalazo committed:
Improved error message from the parser

alalazo committed:
Revert changes on strumpack

alalazo commented:
> Not completely done, but here is a partial review.

@tgamblin :point_up:

tgamblin reviewed (approved):
I went over the rest of the code here and it LGTM.
